tasks/DeviceDispatcher.hpp

```diff
@@ -0,0 +1,93 @@
+#ifndef NMEA2000_DEVICE_DISPATCHER_HPP
+#define NMEA2000_DEVICE_DISPATCHER_HPP
+
+#include <nmea2000/nmea2000Types.hpp>
+#include <nmea2000/Message.hpp>
+#include <rtt/OutputPort.hpp>
+#include <set>
+
+namespace nmea2000 {
+    /** Common implementation of the device filtering implementation
+     * for both CANTask and ActisenseTask
+     */
+    class DeviceDispatcher {
+    public:
+        enum QueryState {
+            QUERY_COMPLETE, //! all devices have been resolved
+            QUERY_TIMED_OUT, //! not all devices have been resolved in the allocated time
+            QUERY_IN_PROGRESS
+        };
+
+    private:
+        struct DynamicPort {
+            DeviceFilter filter;
+            bool needs_resolution = false;
+            uint8_t bus_id = 0;
+            uint32_t pgn;
+
+            RTT::OutputPort<Message>* port = nullptr;
```

Comment:
I think you would make addPorts/removePorts safer if you used a smart pointer here

Reply:
I don't see how. Can you expand ?

Reply:
First, you don't have to worry about deleting the dynamically allocated ports (which I just noticed actually don't do in `removePorts`...). You could also use unique_ptr's custom deleter to not even have to worry about calling removePorts before addPorts.. Just replace them and let the deleter do the removing...

Reply:
Deletion and removePorts have to work hand-in-hand, or things will crash if the unique_ptr is cleared before removePorts is called anyways. So, to me, this simply removes one problem to create others.

Reply:
Maybe.. But the current implementation leaks the ports if they are changed doesn't it?

Reply:
Yes. So ?

[The example in the README](https://github.com/rock-core/drivers-orogen-iodrivers_base) crashes on error, but is still using `unique_ptr` and RAII. Just pointing out that "not leaking" is not "perfect code", and that therefore "leaking" does not mandate "switch to smart pointers".

Reply:
Anyways, fixed the very broken logic of removePort, and cleaned addPort a little.

Reply:
I don't know where the "perfect code" argument came from, we both know there is no such thing. 

And "leaking" does not mandate "smart pointers" but smart pointers definitely makes leaking less likely. To be honest, other than a slight performance hit, I don't see any downsides to using smart pointers, therefore I believe they should be used whenever possible in modern C++ code.

Anyway, the code looks fine now...